### PR TITLE
Added wrapper functions for STRING_AGG() function to fix issue with input containing multi-byte characters.

### PR DESCRIPTION
**Pull request by rohit01010**
### Description
Currently for string function `STRING_AGG()` we are directly using corresponding PG function whose input and output datatype is `TEXT`. This leads to incorrect output for input containing multi-byte characters with `nchar` and `nvarchar` datatype. This PR will fix this issue by adding wrapper functions with correct input and return datatype for `STRING_AGG()` function.

Authored-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)
Signed-off-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)

### Issues Resolved

BABEL-5156

### Test Scenarios Covered ###
* **Use case based -** YES

* **Boundary conditions -** YES

* **Arbitrary inputs -** YES

* **Negative test cases -** YES

* **Minor version upgrade tests -** YES

* **Major version upgrade tests -** YES

* **Performance tests -** NO

* **Tooling impact -** NO

* **Client tests -** NO

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/JDBC/expected/string_agg-vu-verify.out [hunk 1]
```diff
@@ -0,0 +1,1352 @@
...
+DECLARE @sep NVARCHAR(10) = '-'
+SELECT STRING_AGG(col_varchar, @sep) FROM babel_5156_string_agg_t1 GROUP BY group_id ORDER BY group_id
+GO
+~~START~~
+varchar
+abc-?defghi-??
+??·?-??
+<NULL>
+Rohit-Bhagat
+~~END~~
```
Comment (Anikait143) [inline test/JDBC/expected/string_agg-vu-verify.out]: Document mentions that `If the input expression is type VARCHAR, the separator cannot be type NVARCHAR.`, we should handle this case as well

File: test/JDBC/expected/string_agg-vu-verify.out [hunk 2]
```diff
@@ -0,0 +1,1352 @@
...
+
+
+SELECT STRING_AGG(col_nchar, '-') FROM master_dbo.babel_5156_string_agg_t1 GROUP BY group_id ORDER BY group_id;
+GO
+~~START~~
+text
+abc                 -🙂defghi             -🙂🙂                  
+比尔·拉                -莫斯                  
+<NULL>
```
Comment (Anikait143) [inline test/JDBC/expected/string_agg-vu-verify.out:1312]: is this `NULL` expected?

File: test/JDBC/expected/non_default_server_collation/japanese_ci_as/string_agg-before-15_9-or-16_5-vu-verify.out
Comment (Anikait143) [inline test/JDBC/expected/non_default_server_collation/japanese_ci_as/string_agg-before-15_9-or-16_5-vu-verify.out:1]: Please add other verify files in japanese collation folder as well

File: test/JDBC/expected/string_agg-vu-verify.out (hunk 1 above)
Comment (rohit01010) [reply test/JDBC/expected/string_agg-vu-verify.out]: Added check to handle this scenario.

File: test/JDBC/expected/string_agg-vu-verify.out (hunk 2 above)
Comment (rohit01010) [reply test/JDBC/expected/string_agg-vu-verify.out:1312]: Yes, it is expected.

File: test/JDBC/expected/non_default_server_collation/japanese_ci_as/string_agg-before-15_9-or-16_5-vu-verify.out
Comment (rohit01010) [reply test/JDBC/expected/non_default_server_collation/japanese_ci_as/string_agg-before-15_9-or-16_5-vu-verify.out:1]: Added

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 3]
```diff
@@ -301,6 +301,8 @@ tsql_cast_raw_info_t tsql_cast_raw_infos[] =
 	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "sys", "uniqueidentifier", "sys", "bpchar", NULL, 'i', 'i'},
 	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "sys", "uniqueidentifier", "pg_catalog", "varchar", NULL, 'i', 'i'},
 	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "sys", "uniqueidentifier", "sys", "varchar", NULL, 'i', 'i'},
+/*  bit -> string via I/O */
```
Comment (shah-nirmit) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c:304]: Can we add few more tests for this cast exclusively , cause I dont think we have enough coverage here.

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 4]
```diff
@@ -1184,6 +1184,15 @@ class tsqlCommonMutator : public TSqlParserBaseListener
 			rewritten_query_fragment.emplace(std::make_pair(linked_srv->start->getStartIndex(), std::make_pair(linked_srv_name, str)));
 		}	
 	}
+
+	void exitSTRING_AGG(TSqlParser::STRING_AGGContext *ctx) override
```
Comment (shah-nirmit) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: Nit: Please add comment regarding why this change is needed

Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: Added

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 3 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c:304]: Added

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp (hunk 4 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: Upon careful investigation find out that, we do not need to rewrite in `exitSTRING_AGG` function as `exitSTRING_AGG` function will only be called if `STRING_AGG` function is using `WITHIN GROUP` and since `WITHIN GROUP` is supported from Babelfish v16.4 onwards, we do not need to rewrite the `STRING_AGG` here.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 5]
```diff
@@ -1159,6 +1172,38 @@ validate_special_function(char *func_nsname, char *func_name, List* fargs, int n
...
+			|| (*common_utility_plugin_ptr->is_tsql_bpchar_datatype)(input_typeids[0])
+			|| (input_typeids[0] == UNKNOWNOID && !(IsA(first_arg, Const) && ((Const *)first_arg)->constisnull)))
+		{
+			if ((*common_utility_plugin_ptr->is_tsql_nvarchar_datatype)(input_typeids[1])
+				|| (*common_utility_plugin_ptr->is_tsql_nchar_datatype)(input_typeids[1]))
+			{
+				ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_FUNCTION),
+					errmsg("Argument data type %s is invalid for argument %d of %s function.", 
+							format_type_be(input_typeids[1]), 2, special_func->formatted_funcname)));
```
Comment (Anikait143) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c:1202]: This error message seems a bit misleading as `NVARCHAR` is anyway a valid argument type it's just the combination which is not accepted, can we have a more meaningful message

File: test/JDBC/input/functions/string_functions/string_agg-vu-verify.mix
Comment (Anikait143) [inline test/JDBC/input/functions/string_functions/string_agg-vu-verify.mix:1]: Have we added tests where expression  is `NULL` verifying the property: `Null values are ignored and the corresponding separator is not added`?

Comment (rohit01010) [reply test/JDBC/input/functions/string_functions/string_agg-vu-verify.mix:1]: Yes I had added test to cover this scenario. Refer to the table `babel_5156_string_agg_t1`, the values with group_id = 4 are `'Rohit'`, `NULL`, `'Bhagat'`. Now for following query we see can that in the 4th entry first and last value got concatenated, and NULL value is ignored also it's corresponding seperator is not added.
```
SELECT STRING_AGG(col_varchar, '-') FROM babel_5156_string_agg_t1 GROUP BY group_id ORDER BY group_id
GO
~~START~~
varchar
abc-?defghi-??
??·?-??
<NULL>
Rohit-Bhagat
~~END~~
```

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 5 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c:1202]: TSQL throws the same error, and since we want to align with its behaviour used this error message.

**Review by Anikait143 — APPROVED** [review]

**Review by shah-nirmit — APPROVED** [review]
LGTM

**Review by Anikait143 — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 6]
```diff
@@ -1159,6 +1172,38 @@ validate_special_function(char *func_nsname, char *func_name, List* fargs, int n
...
+	{
+		Node *first_arg = (Node *) linitial(fargs);
+		/* if common_utility_plugin_ptr is not initialised */
+		if (common_utility_plugin_ptr == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_INTERNAL_ERROR),
+						errmsg("Failed to find common utility plugin.")));
+
+		/* 
+		 * if first argument if of type CHAR/VARCHAR/STRING_LITERAL and second argument is of type NCHAR/NVARCHAR then throw error.
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: typo: if -> is

**Review by Deepesh125 — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 6 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Updated

**Review by Deepesh125 — APPROVED** [review]
